File: svws-core/src/main/java/de/svws_nrw/core/utils/gost/GostBlockungsergebnisListeneintragComparator.java
package de.svws_nrw.core.utils.gost;

import java.util.Comparator;

import de.svws_nrw.core.data.gost.GostBlockungsergebnisListeneintrag;
import jakarta.validation.constraints.NotNull;

/**
 * Dieser Klasse dient dazu Objekte des Typs {@link GostBlockungsergebnisListeneintrag} zu sortieren.
 */
public final class GostBlockungsergebnisListeneintragComparator implements Comparator<@NotNull GostBlockungsergebnisListeneintrag> {

	/**
	 * Erzeugt einen Comparator für zwei Objekte des Typs {@link GostBlockungsergebnisListeneintrag}.
	 * Zwei Elemente werden nach folgender Priorität sortiert:
	 * <br>(1) Array an Regelverletzungen (weniger besser) + Anzahl nicht gesetzter Kurse
	 * <br>(2) Summe nicht zugeordneter Fachwahlen + Summe an Kollisionen (weniger besser)
	 * <br>(3) Array an Kursdifferenzen-Histogramm (kleinste größte Kursdifferenz besser)
	 * <br>(4) Summe gleicher Facharten in der selben Schiene (weniger besser)
	 * <br>(5) ID von GostBlockungsergebnisListeneintrag (weniger besser)
	 */
	public GostBlockungsergebnisListeneintragComparator() {
		// empty constructor
	}

	@Override
	public int compare(final @NotNull GostBlockungsergebnisListeneintrag o1, final @NotNull GostBlockungsergebnisListeneintrag o2) {
		// Bewertungskriterium 1-4:
		final int cmp = GostBlockungsergebnisBewertungComparator.compareBewertungen(o1.bewertung, o2.bewertung);
		if (cmp != 0)
			return cmp;

		// Bewertungskriterium 5: Bei absolut identischen Ergebnissen gewinnt die kleinere ID des Zwischenergebnisses.
		if (o1.id < o2.id) return -1;
		if (o1.id > o2.id) return +1;

		return 0; // Sollte niemals, da die IDs niemals gleich sind.
	}

}

File: svws-webclient/core/src/core/utils/gost/GostBlockungsergebnisListeneintragComparator.ts
import { JavaObject } from '../../../java/lang/JavaObject';
import { GostBlockungsergebnisBewertungComparator } from '../../../core/utils/gost/GostBlockungsergebnisBewertungComparator';
import { GostBlockungsergebnisListeneintrag } from '../../../core/data/gost/GostBlockungsergebnisListeneintrag';
import type { Comparator } from '../../../java/util/Comparator';

export class GostBlockungsergebnisListeneintragComparator extends JavaObject implements Comparator<GostBlockungsergebnisListeneintrag> {


	/**
	 * Erzeugt einen Comparator für zwei Objekte des Typs {@link GostBlockungsergebnisListeneintrag}.
	 * Zwei Elemente werden nach folgender Priorität sortiert:
	 * <br>(1) Array an Regelverletzungen (weniger besser) + Anzahl nicht gesetzter Kurse
	 * <br>(2) Summe nicht zugeordneter Fachwahlen + Summe an Kollisionen (weniger besser)
	 * <br>(3) Array an Kursdifferenzen-Histogramm (kleinste größte Kursdifferenz besser)
	 * <br>(4) Summe gleicher Facharten in der selben Schiene (weniger besser)
	 * <br>(5) ID von GostBlockungsergebnisListeneintrag (weniger besser)
	 */
	public constructor() {
		super();
	}

	public compare(o1 : GostBlockungsergebnisListeneintrag, o2 : GostBlockungsergebnisListeneintrag) : number {
		const cmp : number = GostBlockungsergebnisBewertungComparator.compareBewertungen(o1.bewertung, o2.bewertung);
		if (cmp !== 0)
			return cmp;
		if (o1.id < o2.id)
			return -1;
		if (o1.id > o2.id)
			return +1;
		return 0;
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.core.utils.gost.GostBlockungsergebnisListeneintragComparator';
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.utils.gost.GostBlockungsergebnisListeneintragComparator', 'java.util.Comparator'].includes(name);
	}

}

export function cast_de_svws_nrw_core_utils_gost_GostBlockungsergebnisListeneintragComparator(obj : unknown) : GostBlockungsergebnisListeneintragComparator {
	return obj as GostBlockungsergebnisListeneintragComparator;
}
